odc/geo/cog/_tifffile.py

```diff
@@ -573,6 +583,33 @@ def _gdal_level(compression: str, default=None) -> Any:
     return (predictor, compression, compressionargs)
 
 
+def _gdal_sample_description(sample: int, description: str) -> str:
+    """Make XML line of GDAL metadata.
+
+    :param band: Sample / band number in data array.
+    :param description: Band name in data array.
+
+    :return: GDAL XML metadata line to place in TIFF file.
+    """
+    # GDAL does double escaping; see frmts/gtiff/geotiff.cpp.
+    # We also double escape to maximize compatibility with tools expecting GDAL-generated metadata.
+    double_escaped_description = xml_escape(xml_escape(description))
+    return f'<Item name="DESCRIPTION" sample="{sample}" role="description">{double_escaped_description}</Item>'
+
+
+def _gdal_sample_descriptions(xx: xr.DataArray) -> List[str]:
+    """Translate ``long_name`` attribute (if present) to GDAL sample description metadata.
+
+    :param xx: Pixels as :py:class:`xarray.DataArray` backed by Dask
+
+    :return: List of GDAL XML metadata lines to place in TIFF file.
+    """
+    return [
+        _gdal_sample_description(sample, description)
+        for sample, description in enumerate(xx.attrs.get("long_name", []))
+    ]
```

Comment:
Can we make this a bit more flexible on input?

1. Treat `long_name="something"` as equivalent to `long_name=["something"]`
2. If there is `band` coordinate containing strings, use that as the source metadata
3. If there is `long_name` coordinate, use that

It's common to start off with data like `ds: xarray.Dataset` then use something like `xx = ds.to_array('band')` before saving that to multi-sample COG, would be nice if that generated descriptions with original data variable names.

Other note: While I do appreciate documentation comments even on internal functions, function name and type annotations are often enough.

Reply:
I added a `_band_names` function above that takes care of all of this.  We now use the `band` coordinate if it's present and it's a string; otherwise we fall back to the `long_name` attribute, and that can be a single string or a list of strings.